File: solutions/ruby/raindrops/7/raindrops.rb
class Raindrops
  attr_reader :number, :sound

  def initialize(number)
    @number = number
    FACTOR_NOISES.each_with_object(@sound = '') do |(factor, noise), sound|
      sound << noise if factor?(number, factor)
    end
  end

  def self.convert(number)
    new(number).to_s
  end

  def to_s
    (sound.empty? ? number : sound).to_s
  end

  private

  FACTOR_NOISES = {
    3 => 'Pling',
    5 => 'Plang',
    7 => 'Plong'
  }.freeze

  def factor?(number, divisor)
    (number % divisor).zero?
  end
end

File: solutions/python/raindrops/1/raindrops.py
"""Module to convert a number to a raindrop sound"""

def convert(number: int) -> str:
    """
    Convert an integer number to a raindrop sound.

    :param number: int - the number to convert
    :return: str - thr raindrop sound
    """
    
    sound_tests: list[tuple[int, str]] = [
        (3, "Pling"),
        (5, "Plang"),
        (7, "Plong")
    ]

    sound = "".join(test[1] for test in sound_tests if number % test[0] == 0)

    return sound if sound else str(number)
